server/container/util/src/main/java/org/apache/james/util/MemoizedSupplier.java

```diff
@@ -19,12 +19,38 @@
 
 package org.apache.james.util;
 
+import java.util.function.Consumer;
 import java.util.function.Supplier;
 
 import com.google.common.base.Suppliers;
 
-public class MemoizedSupplier {
-    public static <T> Supplier<T> of(Supplier<T> originalSupplier) {
-        return Suppliers.memoize(originalSupplier::get)::get;
+public class MemoizedSupplier<T> implements Supplier<T> {
```

Comment:
Ok for me, but maybe we should:
 - Add a comment explaining this class is taken from guava (+ version) and which changes we did to it.
 - If this class is tested in Guava, we need to add its tests too. Otherwise we need to write it IMO...

Reply:
We have can leverage old test class to do test for this new `MemoizedSupplier`. I think I just need to write additional tests for `ifInitialized()`.